tests/transforms/test_decompose_transform.py

```diff
@@ -0,0 +1,240 @@
+        ),
+        (
+            [qml.Toffoli([0, 1, 2])],
+            {qml.Toffoli},
+            [qml.Toffoli([0, 1, 2])],
+            None,
+        ),
+        (
+            [qml.measurements.MidMeasureMP(0)],
+            {},
+            [qml.measurements.MidMeasureMP(0)],
+            "MidMeasureMP",
+        ),
+    ]
+
+    callables_test = [
+        (
+            [qml.Hadamard(0)],
+            lambda op: "RX" in op.name,
+            [qml.RZ(qnp.pi / 2, 0), qml.RX(qnp.pi / 2, 0), qml.RZ(qnp.pi / 2, 0)],
+            "RZ",
+        ),
+        (
+            [qml.Toffoli([0, 1, 2])],
+            lambda op: len(op.wires) <= 2,
+            [
+                qml.Hadamard(wires=[2]),
+                qml.CNOT(wires=[1, 2]),
+                qml.ops.op_math.Adjoint(qml.T(wires=[2])),
+                qml.CNOT(wires=[0, 2]),
+                qml.T(wires=[2]),
+                qml.CNOT(wires=[1, 2]),
+                qml.ops.op_math.Adjoint(qml.T(wires=[2])),
+                qml.CNOT(wires=[0, 2]),
+                qml.T(wires=[1]),
+                qml.T(wires=[2]),
+                qml.CNOT(wires=[0, 1]),
+                qml.Hadamard(wires=[2]),
+                qml.T(wires=[0]),
+                qml.ops.op_math.Adjoint(qml.T(wires=[1])),
+                qml.CNOT(wires=[0, 1]),
+            ],
+            None,
+        ),
+    ]
+
+    @pytest.mark.parametrize("gate_set", gate_set_inputs)
+    def test_different_input_formats(self, gate_set):
+        """Tests that gate sets of different types are handled correctly"""
+        tape = qml.tape.QuantumScript([qml.RX(0, wires=[0])])
+        decompose(tape, gate_set=gate_set)
+
+    def test_user_warning(self):
+        """Tests that user warning is raised if operator does not have a valid decomposition"""
+        tape = qml.tape.QuantumScript([qml.RX(0, wires=[0])])
+        with pytest.warns(UserWarning, match="has no supported decomposition"):
+            decompose(tape, gate_set=lambda op: op.name not in {"RX"})
+
+    def test_infinite_decomposition_loop(self):
+        """Test that a recursion error is raised if decomposition enters an infinite loop."""
+        tape = qml.tape.QuantumScript([InfiniteOp(1.23, 0)])
+        with pytest.raises(RecursionError, match=r"Reached recursion limit trying to decompose"):
+            decompose(tape, lambda obj: obj.has_matrix)
+
+    @pytest.mark.parametrize("initial_ops, gate_set, expected_ops, warning_pattern", iterables_test)
+    def test_iterable_gate_set(self, initial_ops, gate_set, expected_ops, warning_pattern):
+        """Tests that gate sets defined with iterables decompose correctly"""
+        tape = qml.tape.QuantumScript(initial_ops)
+
+        if warning_pattern is not None:
+            with pytest.warns(UserWarning, match=warning_pattern):
+                (decomposed_tape,), _ = decompose(tape, gate_set=gate_set)
+        else:
+            (decomposed_tape,), _ = decompose(tape, gate_set=gate_set)
```

Comment:
```suggestion
            with warnings.catch_warnings():
                warnings.simplefilter("error")
                (decomposed_tape,), _ = decompose(tape, gate_set=gate_set)
```
(hopefully spaces are in the right place).

Might be good to also make sure we are not raising warnings when we don't want to be raising warnings.

Potentially we could also add a fixture to the entire file:

```
@pytest.fixture(autouse=True)
def warnings_as_errors():
    with warnings.catch_warnings():
        warnings.simplefilter("error")
        yield
```

So that we are making sure that we are never raising warnings when we don't explicitly want to be raising warnings.

Reply:
Nice! I like the fixture idea, very clean. 👍🏼